// @pugdag/wallet/dist/wallet/address-manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressManager = void 0;
// @ts-ignore
const pugdagcore = __importStar(require("../../../core-lib"));
// @ts-ignore
const secp256k1 = pugdagcore.secp256k1; //require('secp256k1-wasm');
const event_target_impl_1 = require("./event-target-impl");
const helper_1 = require("../utils/helper");
class AddressManager extends event_target_impl_1.EventTargetImpl {
    constructor(HDWallet, network) {
        super();
        /**
         * Derives a new receive address. Sets related instance properties.
         */
        this.receiveAddress = {
            counter: 0,
            // @ts-ignore
            current: {},
            keypairs: {},
            atIndex: {},
            next: () => {
                const { address, privateKey } = this.deriveAddress('receive', this.receiveAddress.counter);
                this.receiveAddress.current = {
                    address,
                    privateKey
                };
                this.receiveAddress.keypairs[address] = privateKey;
                this.receiveAddress.atIndex[this.receiveAddress.counter] = address;
                this.receiveAddress.counter += 1;
                return address;
            },
            advance(n) {
                if (n > -1)
                    this.counter = n;
                this.next();
            },
        };
        /**
         * Derives a new change address. Sets related instance properties.
         */
        this.changeAddress = {
            counter: 0,
            // @ts-ignore
            current: {},
            keypairs: {},
            atIndex: {},
            next: () => {
                const { address, privateKey } = this.deriveAddress('change', this.changeAddress.counter);
                this.changeAddress.keypairs[address] = privateKey;
                this.changeAddress.current = {
                    address,
                    privateKey
                };
                this.changeAddress.atIndex[this.changeAddress.counter] = address;
                this.changeAddress.counter += 1;
                return address;
            },
            advance(n) {
                if (n > -1)
                    this.counter = n;
                // no call to next() here; composeTx calls it on demand.
            },
            reverse() {
                if (this.counter > 0)
                    this.counter -= 1;
            },
        };
        this.HDWallet = HDWallet;
        this.network = network;
    }
    get all() {
        return Object.assign(Object.assign({}, this.receiveAddress.keypairs), this.changeAddress.keypairs);
    }
    get shouldFetch() {
        const receive = Object.entries(this.receiveAddress.atIndex)
            .filter((record) => parseInt(record[0], 10) <= this.receiveAddress.counter - 1)
            .map((record) => record[1]);
        const change = Object.entries(this.changeAddress.atIndex)
            .filter((record) => parseInt(record[0], 10) <= this.changeAddress.counter)
            .map((record) => record[1]);
        return [...receive, ...change];
    }
    deriveAddress(deriveType, index) {
        //let ts0 = Date.now();
        const dType = deriveType === 'receive' ? 0 : 1;
        const { privateKey } = this.HDWallet.deriveChild(`m/44'/972/0'/${dType}'/${index}'`);
        //let ts1 = Date.now();
        //let publicKeys = secp256k1.export_public_keys(privateKey.toString());
        const xonlyPubKey = secp256k1.export_public_key_xonly(privateKey.toString());
        //let ts2 = Date.now();
        //console.log('durations:',(ts2-ts1)/1000,(ts1-ts0)/1000);
        //let address1 = new pugdagcore.PublicKey(publicKeys.pubkey, {network:this.network}).toAddress().toString();
        //let address = privateKey.toAddress(this.network).toString();
        //let pubkey = Buffer.from(publicKeys.pubkey, "hex");
        //let {address:address3} = bitcoin.payments.p2pkh({pubkey});
        let xonly = Buffer.from(xonlyPubKey, "hex");
        //@ts-ignore
        let address = pugdagcore.Address.fromPublicKeyBuffer(xonly, this.network).toString();
        /*
        console.log("privateKey:xxxx:", {
          privateKey: privateKey.toString(),
          address,
          address1,
          address2,
          "address1==address":address1==address,
          publicKeys
         });//, publicKeys)
         */
        //console.log("xonly:address2", "privateKey:"+privateKey.toString(), "address:"+address2)
        //console.log("xonly", publicKeys.xonly)
        (0, helper_1.dpc)(() => {
            this.emit("new-address", {
                type: deriveType,
                address,
                index
            });
        });
        return {
            address,
            privateKey
        };
    }
    /**
     * Derives n addresses and adds their keypairs to their deriveType-respective address object
     * @param n How many addresses to derive
     * @param deriveType receive or change address
     * @param offset Index to start at in derive path
     */
    getAddresses(n, deriveType, offset = 0) {
        return [...Array(n).keys()].map((i) => {
            const index = i + offset;
            const { address, privateKey } = this.deriveAddress(deriveType, index);
            if (deriveType === 'receive') {
                this.receiveAddress.atIndex[index] = address;
                this.receiveAddress.keypairs[address] = privateKey;
            }
            else {
                this.changeAddress.atIndex[index] = address;
                this.changeAddress.keypairs[address] = privateKey;
            }
            return {
                index,
                address,
                privateKey,
            };
        });
    }
    isOur(address) {
        return !!(this.changeAddress.keypairs[address] || this.receiveAddress.keypairs[address]);
    }
    isOurChange(address) {
        return !!this.changeAddress.keypairs[address];
    }
}
exports.AddressManager = AddressManager;
