// @pugdag/wallet/dist/wallet/api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PugdagAPI = exports.ApiError = void 0;
const event_target_impl_1 = require("./event-target-impl");
class ApiError {
    constructor(message, debugInfo = null) {
        //super(...args);
        this.name = 'ApiError';
        this.message = message;
        this.debugInfo = debugInfo;
        if (!Error.captureStackTrace)
            this.stack = ((new Error(message)).stack + "").split("↵").join("\n");
        else
            Error.captureStackTrace(this, ApiError);
    }
    setDebugInfo(info) {
        this.debugInfo = info;
    }
    setExtraDebugInfo(info) {
        this.extraDebugInfo = info;
    }
}
exports.ApiError = ApiError;
const missingRPCProviderError = () => {
    throw new ApiError(`RPC privider is missing. Please set RPC using 
		Wallet.setRPC(rpc_provider).`);
};
class PugdagAPI extends event_target_impl_1.EventTargetImpl {
    constructor() {
        super(...arguments);
        this.isConnected = false;
    }
    // constructor(rpc:IRPC) {
    // 	this.rpc = rpc;
    // }
    setRPC(rpc) {
        this.rpc = rpc;
        rpc.onConnect(() => {
            this._setConnected(true);
        });
        rpc.onDisconnect(() => {
            this._setConnected(false);
        });
    }
    getRPC() {
        // @ts-ignore
        return this.rpc;
    }
    on(type, callback) {
        super.on(type, callback);
        if (type == 'connect' && this.isConnected) {
            //console.log("api.on->connect", this.isConnected, callback)
            callback({}, { type, detail: {}, defaultPrevented: false });
        }
    }
    _setConnected(isConnected) {
        //console.log("wallet.api._setConnected", isConnected)
        this.isConnected = isConnected;
        this.emit(isConnected ? "connect" : 'disconnect');
    }
    buildUtxoMap(entries) {
        let result = new Map();
        entries.map(entry => {
            //console.log("entry", entry)
            let { transactionId, index } = entry.outpoint;
            let { address, utxoEntry } = entry;
            let { amount, scriptPublicKey, blockDaaScore, isCoinbase } = utxoEntry;
            let item = {
                amount,
                scriptPublicKey,
                blockDaaScore,
                transactionId,
                index,
                isCoinbase
            };
            let items = result.get(address);
            if (!items) {
                items = [];
                result.set(address, items);
            }
            items.push(item);
        });
        return result;
    }
    buildOutpointMap(outpoints) {
        const map = new Map();
        outpoints.map(item => {
            let list = map.get(item.address);
            if (!list) {
                list = [];
                map.set(item.address, list);
            }
            list.push(item.outpoint);
        });
        return map;
    }
    getVirtualSelectedParentBlueScore() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.getVirtualSelectedParentBlueScore()
                .catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return { blueScore: response.blueScore };
        });
    }
    getVirtualDaaScore() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.getBlockDagInfo()
                .catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return { virtualDaaScore: response.virtualDaaScore };
        });
    }
    subscribeVirtualSelectedParentBlueScoreChanged(callback) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.subscribeVirtualSelectedParentBlueScoreChanged(callback)
                .catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return response;
        });
    }
    subscribeVirtualDaaScoreChanged(callback) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.subscribeVirtualDaaScoreChanged(callback)
                .catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return response;
        });
    }
    subscribeUtxosChanged(addresses, callback) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const cb = (res) => {
                // console.log("UtxosChangedNotification:", res)
                const added = this.buildUtxoMap(res.added);
                const removed = this.buildOutpointMap(res.removed);
                callback(added, removed);
            };
            let p = this.rpc.subscribeUtxosChanged(addresses, cb);
            let { _utxosChangedSubUid } = this;
            let { uid } = p;
            this._utxosChangedSubUid = uid;
            const response = yield p.catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (_utxosChangedSubUid)
                this.rpc.unSubscribeUtxosChanged(_utxosChangedSubUid);
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return response;
        });
    }
    getUtxosByAddresses(addresses) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.getUtxosByAddresses(addresses).catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return this.buildUtxoMap(response.entries);
        });
    }
    submitTransaction(tx) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            // eslint-disable-next-line
            const response = yield this.rpc.submitTransaction(tx).catch((e) => {
                throw new ApiError(`API connection error. ${e}`); // eslint-disable-line
            });
            //console.log("submitTransaction:result", response)
            if (response.transactionId)
                return response.transactionId;
            if (!response.error)
                response.error = { message: 'Api error. Please try again later. (ERROR: SUBMIT-TX:100)' };
            if (!response.error.errorCode)
                response.error.errorCode = 100;
            throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`, tx);
        });
    }
    getBlock(blockHash) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            // eslint-disable-next-line
            const response = yield this.rpc.getBlock(blockHash).catch((e) => {
                throw new ApiError(`API connection error. ${e}`); // eslint-disable-line
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            return response.blockVerboseData;
        });
    }
    ;
    // TODO: handle pagination
    getTransactionsByAddresses(addresses, startingBlockHash = "") {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rpc)
                return missingRPCProviderError();
            const response = yield this.rpc.getTransactionsByAddresses(startingBlockHash, addresses).catch((e) => {
                throw new ApiError(`API connection error. ${e}`);
            });
            if (response.error)
                throw new ApiError(`API error (${response.error.errorCode}): ${response.error.message}`);
            let { transactions, lasBlockScanned } = response;
            return { transactions, lasBlockScanned };
        });
    }
}
exports.PugdagAPI = PugdagAPI;
