// @pugdag/wallet/dist/wallet/cache-store.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheStore = void 0;
const indexed_db_1 = require("./indexed-db");
const tx_store_1 = require("./tx-store");
class CacheStore {
    constructor(wallet) {
        this.store = new Map();
        this.wallet = wallet;
        let { uid, network } = wallet;
        console.log("CacheStore:wallet:uid", uid);
        let sNetwork = tx_store_1.internalNames[network] || network;
        if (typeof indexedDB != "undefined")
            this.idb = new indexed_db_1.iDB({ storeName: "cache", dbName: "pugdag_" + uid + "_" + sNetwork });
    }
    setAddressIndexes(data) {
        let item = Object.assign({
            id: "address-indexes",
            ts: Date.now()
        }, data);
        this.set(item);
    }
    getAddressIndexes() {
        return this.get("address-indexes");
    }
    set(item, skipSave = false) {
        this.store.set(item.id, item);
        this.emitCache(item);
        if (!skipSave)
            this.save(item);
    }
    get(id) {
        return this.store.get(id);
    }
    save(item) {
        var _a;
        (_a = this.idb) === null || _a === void 0 ? void 0 : _a.set(item.id, JSON.stringify(item));
    }
    emitCache(item) {
        this.wallet.emit("wallet-cache", item);
    }
    restore() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.idb) {
                let entries = (yield this.idb.entries().catch((err) => {
                    console.log("cache-store: entries():error", err);
                })) || [];
                let length = entries.length;
                console.log("cache idb entries:", entries);
                let list = [];
                for (let i = 0; i < length; i++) {
                    let [key, cacheStr] = entries[i];
                    if (!cacheStr)
                        continue;
                    try {
                        let cacheItem = JSON.parse(cacheStr);
                        list.push(cacheItem);
                    }
                    catch (e) {
                        this.wallet.logger.error("CACHE parse error - 104:", cacheStr, e);
                    }
                }
                list.sort((a, b) => {
                    return a.ts - b.ts;
                }).map(o => {
                    this.set(o, true);
                });
            }
        });
    }
}
exports.CacheStore = CacheStore;
