// @pugdag/wallet/dist/wallet/tx-store.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TXStore = exports.internalNames = void 0;
const indexed_db_1 = require("./indexed-db");
const API_BASE = "https://api.pugdagcoin.com/";
exports.internalNames = {
    mainnet: "default",
    pugdag: "default",
    testnet: "testnet",
    pugdagtest: "testnet",
    pugdagsim: "simnet",
    pugdagdev: "devnet",
    pugdagreg: "pugdagreg"
};
class TXStore {
    constructor(wallet) {
        this.store = new Map();
        this.txToEmitList = [];
        this.updatedTxToEmitList = [];
        this.pendingUpdate = [];
        this.updateTxTimeoutId = null;
        this.emitTxTimeoutId = null;
        this.emitUpdateTxTimeoutId = null;
        this.updatingTransactionsInprogress = false;
        this.transactionUpdating = false;
        this.wallet = wallet;
        let { uid, network } = wallet;
        let sNetwork = exports.internalNames[network] || network;
        //this.restore();
        if (typeof indexedDB != "undefined")
            this.idb = new indexed_db_1.iDB({ storeName: "tx", dbName: "pugdag_" + uid + "_" + sNetwork });
    }
    add(tx, skipSave = false) {
        //console.log("idb add:tx:", "ts:"+tx.ts, "skipSave:"+skipSave, tx)
        if (this.store.has(tx.id))
            return false;
        this.store.set(tx.id, tx);
        this.emitTx(tx);
        if (this.store.size > TXStore.MAX)
            this.store = new Map([...this.store.entries()].slice(-TXStore.MAX));
        if (!skipSave)
            this.save(tx);
        return true;
    }
    removePendingUTXO(utxo, address = '') {
        let id = utxo.transactionId + ":" + utxo.index;
        let dbItem = this.store.get(id);
        if (dbItem) {
            dbItem.isMoved = true;
            this.store.set(id, dbItem);
            this.save(dbItem);
        }
        else {
            dbItem = {
                in: true,
                ts: Date.now(),
                id,
                amount: utxo.amount,
                address,
                blueScore: utxo.blockDaaScore,
                tx: false, //TODO
                isMoved: true,
                isCoinbase: false
            };
        }
        this.emitTx(dbItem);
    }
    fetchTransactions(txIds) {
        return fetch(`${API_BASE}transactions/search?fields=transaction_id%2Cblock_time`, {
            headers: {
                'Access-Control-Allow-Origin': '*',
                'content-type': 'application/json'
            },
            method: "POST",
            body: JSON.stringify({ "transactionIds": txIds })
        })
            .catch(err => {
            this.wallet.logger.debug("ExplorerAPI transactions/search : error", err);
        })
            .then((response) => {
            this.wallet.logger.debug("ExplorerAPI transactions/search, txIds:", txIds, "Response:", response);
            if (response) {
                return response.json();
            }
        })
            .then(data => {
            this.wallet.logger.debug("ExplorerAPI transactions/search, data:", data);
            if (Array.isArray(data))
                return data;
            return [];
        });
    }
    fetchTxTime(txIds) {
        return __awaiter(this, void 0, void 0, function* () {
            let txs = yield this.fetchTransactions(txIds);
            //this.wallet.logger.info("fetchTransactions: result", txs);
            let txid2time = {};
            if (Array.isArray(txs)) {
                txs.forEach(tx => {
                    txid2time[tx.transaction_id] = tx.block_time;
                });
            }
            return txid2time;
        });
    }
    addAddressUTXOs(address, utxos, ts) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!utxos.length || this.wallet.addressManager.isOurChange(address))
                return;
            utxos.forEach(utxo => {
                let item = {
                    in: true,
                    ts: ts || Date.now(),
                    id: utxo.transactionId + ":" + utxo.index,
                    amount: utxo.amount,
                    address,
                    blueScore: utxo.blockDaaScore,
                    isCoinbase: utxo.isCoinbase,
                    tx: false //TODO
                };
                this.add(item);
            });
        });
    }
    addFromUTXOs(list) {
        let ts = Date.now();
        list.forEach((utxos, address) => {
            this.addAddressUTXOs(address, utxos, ts);
        });
    }
    save(tx) {
        var _a;
        if (this.wallet.options.updateTxTimes) {
            this.updateTransactionTime(tx.id);
        }
        if (typeof indexedDB != "undefined") {
            (_a = this.idb) === null || _a === void 0 ? void 0 : _a.set(tx.id, JSON.stringify(tx));
        }
    }
    updateTransactionTime(id) {
        this.wallet.logger.debug("updateTransactionTime", id);
        this.pendingUpdate.push(id);
        if (this.updateTxTimeoutId) {
            clearTimeout(this.updateTxTimeoutId);
        }
        if (this.pendingUpdate.length > 500) {
            this.updateTransactionTimeImpl();
        }
        else {
            this.updateTxTimeoutId = setTimeout(() => {
                this.updateTransactionTimeImpl();
            }, 10000);
        }
    }
    emitTx(tx) {
        if (this.wallet.syncSignal && !this.wallet.syncInProggress) {
            if (tx.isMoved) {
                this.wallet.emit("moved-transaction", tx);
            }
            else {
                this.wallet.emit("new-transaction", tx);
            }
            return;
        }
        if (this.emitTxTimeoutId) {
            clearTimeout(this.emitTxTimeoutId);
        }
        this.txToEmitList.push(tx);
        if (this.txToEmitList.length > 500) {
            this.emitTxs();
        }
        else {
            this.emitTxTimeoutId = setTimeout(() => {
                this.emitTxs();
            }, 3000);
        }
    }
    emitTxs() {
        let list = this.txToEmitList;
        this.txToEmitList = [];
        this.wallet.emit("transactions", list);
    }
    emitUpdateTx(tx) {
        this.updatedTxToEmitList.push(tx);
        if (this.emitUpdateTxTimeoutId) {
            clearTimeout(this.emitUpdateTxTimeoutId);
        }
        if (this.updatedTxToEmitList.length > 500) {
            this.emitUpdateTxImpl();
        }
        else {
            this.emitUpdateTxTimeoutId = setTimeout(() => {
                this.emitUpdateTxImpl();
            }, 3000);
        }
    }
    emitUpdateTxImpl() {
        let list = this.updatedTxToEmitList;
        this.updatedTxToEmitList = [];
        this.wallet.emit("update-transactions", list);
    }
    startUpdatingTransactions(version = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            this.wallet.logger.info("startUpdatingTransactions:", this.updatingTransactionsInprogress);
            if (this.updatingTransactionsInprogress) {
                this.wallet.emit("transactions-update-status", { status: "in-progress" });
                return false;
            }
            let { txWithMissingVersion: ids } = yield this.getDBEntries(version);
            if (ids.length) {
                this.updatingTransactionsInprogress = true;
                this.wallet.emit("transactions-update-status", { status: "started" });
                yield this.updateTransactionTimeImpl(ids, true, () => {
                    this.updatingTransactionsInprogress = false;
                    this.wallet.emit("transactions-update-status", { status: "finished" });
                });
            }
            else {
                this.wallet.emit("transactions-update-status", { status: "finished", total: 0, updated: 0 });
            }
            return true;
        });
    }
    updateTransactionTimeImpl(txIdList = null, notify = false, callback = null) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.transactionUpdating) {
                setTimeout(() => {
                    this.updateTransactionTimeImpl(txIdList, notify, callback);
                }, 2000);
                return;
            }
            this.transactionUpdating = true;
            let ids = txIdList || this.pendingUpdate;
            let total = 0;
            this.pendingUpdate = [];
            this.wallet.logger.debug("updateTransactionTimeImpl:", ids);
            const CHUNK_SIZE = 500;
            let chunks = [];
            let txIds = [];
            let txId2Id = {};
            ids.map(id => {
                let txId = id.split(":")[0];
                if (!txId2Id[txId]) {
                    txId2Id[txId] = [];
                    txIds.push(txId);
                    total++;
                    if (txIds.length == CHUNK_SIZE) {
                        chunks.push(txIds);
                        txIds = [];
                    }
                }
                txId2Id[txId].push(id);
            });
            if (notify) {
                this.wallet.emit("transactions-update-status", {
                    status: "progress",
                    total,
                    updated: 0
                });
            }
            if (txIds.length) {
                chunks.push(txIds);
            }
            const updateTx = (id, ts = 0) => __awaiter(this, void 0, void 0, function* () {
                let tx = null;
                if (this.idb) {
                    let txStr = yield this.idb.get(id);
                    try {
                        tx = JSON.parse(txStr);
                    }
                    catch (e) {
                        tx = {};
                    }
                }
                tx = tx || {};
                if (ts) {
                    tx.ts = ts;
                    tx.version = 2;
                }
                else {
                    tx.version = 1;
                }
                if (tx.id == id && this.idb) {
                    this.idb.set(id, JSON.stringify(tx));
                }
                tx.id = id;
                this.emitUpdateTx(tx);
                this.wallet.logger.debug("updateTransactionTimeImpl: tx updated", id, "ts:", ts, tx);
            });
            let updatedCount = 0;
            let fetch_txs = () => __awaiter(this, void 0, void 0, function* () {
                let txIds = chunks.shift();
                //this.wallet.logger.info("updateTransactionTimeImpl: fetch_txs", txIds);
                if (!txIds) {
                    this.transactionUpdating = false;
                    callback === null || callback === void 0 ? void 0 : callback();
                    return;
                }
                let count = txIds.length;
                let txId2time = yield this.fetchTxTime(txIds);
                //this.wallet.logger.info("updateTransactionTimeImpl: txId2time", txId2time);
                Object.keys(txId2time).forEach(txId => {
                    let ts = txId2time[txId];
                    let index = txIds.indexOf(txId);
                    if (index > -1) {
                        txIds.splice(index, 1);
                    }
                    txId2Id[txId].forEach((id) => __awaiter(this, void 0, void 0, function* () {
                        yield updateTx(id, ts);
                    }));
                });
                //txs which failed to fetch
                if (this.idb) {
                    txIds.map(txId => {
                        txId2Id[txId].forEach((id) => __awaiter(this, void 0, void 0, function* () {
                            yield updateTx(id);
                        }));
                    });
                }
                updatedCount += count;
                if (notify) {
                    this.wallet.emit("transactions-update-status", {
                        status: "progress",
                        total,
                        updated: updatedCount
                    });
                }
                setTimeout(fetch_txs, 2000);
            });
            setTimeout(fetch_txs, 1000);
        });
    }
    getDBEntries(version = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.idb) {
                return {
                    list: [],
                    txWithMissingVersion: []
                };
            }
            let entries = (yield this.idb.entries().catch((err) => {
                console.log("tx-store: entries():error", err);
            })) || [];
            let length = entries.length;
            console.log("tx-entries length:", length);
            let list = [];
            let ids = [];
            for (let i = 0; i < length; i++) {
                let [key, txStr] = entries[i];
                if (!txStr)
                    continue;
                try {
                    let tx = JSON.parse(txStr);
                    if (tx.version === undefined || (version && tx.version != version)) {
                        ids.push(tx.id);
                    }
                    list.push(tx);
                }
                catch (e) {
                    this.wallet.logger.error("LS-TX parse error - 104:", txStr, e);
                }
            }
            return {
                list,
                txWithMissingVersion: ids
            };
        });
    }
    restore() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.idb) {
                let { list } = yield this.getDBEntries();
                list.sort((a, b) => {
                    return a.ts - b.ts;
                }).map(o => {
                    this.add(o, true);
                });
            }
        });
    }
}
exports.TXStore = TXStore;
TXStore.MAX = 20000;
