// @pugdag/wallet/dist/wallet/unspent-output.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnspentOutput = void 0;
const pugdagcore = __importStar(require("../../../core-lib"));
class UnspentOutput extends pugdagcore.Transaction.UnspentOutput {
    constructor(o) {
        super(o);
        this.blockDaaScore = o.blockDaaScore;
        this.scriptPublicKeyVersion = o.scriptPublicKeyVersion;
        this.id = this.txId + this.outputIndex;
        this.signatureOPCount = this.script.getSignatureOperationsCount();
        this.mass = this.signatureOPCount * pugdagcore.Transaction.MassPerSigOp;
        this.mass += 151 * pugdagcore.Transaction.MassPerTxByte; //standalone mass 
        this.isCoinbase = o.isCoinbase,
            this.scriptPubKey = o.scriptPubKey;
    }
}
exports.UnspentOutput = UnspentOutput;
