// @pugdag/wallet/dist/wallet/utxo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtxoSet = exports.COINBASE_CFM_COUNT = exports.CONFIRMATION_COUNT = exports.UnspentOutput = void 0;
const unspent_output_1 = require("./unspent-output");
Object.defineProperty(exports, "UnspentOutput", { enumerable: true, get: function () { return unspent_output_1.UnspentOutput; } });
const helper = __importStar(require("../utils/helper"));
// import * as api from './apiHelpers';
const wallet_1 = require("./wallet");
const event_target_impl_1 = require("./event-target-impl");
const KLS = helper.KLS;
exports.CONFIRMATION_COUNT = 10;
exports.COINBASE_CFM_COUNT = 100;
let seq = 0;
class UtxoSet extends event_target_impl_1.EventTargetImpl {
    constructor(wallet) {
        super();
        this.utxos = {
            confirmed: new Map(),
            pending: new Map(),
            used: new Map()
        };
        this.inUse = [];
        this.totalBalance = 0;
        this.availableBalance = 0;
        this.debug = false;
        this.utxoStorage = {};
        this.addressesUtxoSyncStatuses = new Map();
        this.wallet = wallet;
    }
    /**
     * Add UTXOs to UTXO set.
     * @param utxos Array of UTXOs from pugdag API.
     * @param address Address of UTXO owner.
     */
    add(utxos, address) {
        const utxoIds = [];
        this.logger.utxodebug("add utxos", utxos);
        const { blueScore } = this.wallet;
        utxos.forEach((utxo) => {
            const utxoId = utxo.transactionId + utxo.index.toString();
            const utxoInUse = this.inUse.includes(utxoId);
            const alreadyHaveIt = !!(this.utxos.confirmed.has(utxoId) || this.utxos.pending.has(utxoId));
            //console.log("utxo.scriptPubKey", utxo)
            //console.log("utxoInUse", {utxoInUse, alreadyHaveIt})
            if (!utxoInUse && !alreadyHaveIt /*&& utxo.isSpendable*/) {
                utxoIds.push(utxoId);
                let confirmed = (blueScore - utxo.blockDaaScore >= (utxo.isCoinbase ? exports.COINBASE_CFM_COUNT : exports.CONFIRMATION_COUNT));
                let unspentOutput = new unspent_output_1.UnspentOutput({
                    txid: utxo.transactionId,
                    address,
                    vout: utxo.index,
                    scriptPubKey: utxo.scriptPublicKey.scriptPublicKey,
                    scriptPublicKeyVersion: utxo.scriptPublicKey.version,
                    satoshis: +utxo.amount,
                    blockDaaScore: utxo.blockDaaScore,
                    isCoinbase: utxo.isCoinbase
                });
                //confirmed = confirmed || this.isOurChange(unspentOutput);
                //confirmed = /*confirmed || */this.isOurChange(unspentOutput);
                //if(confirmed){
                //	console.log("Change address: unspentOutput", blueScore-utxo.blockDaaScore, unspentOutput)
                //}
                let map = this.utxos[confirmed ? 'confirmed' : 'pending'];
                map.set(utxoId, unspentOutput);
                this.wallet.adjustBalance(confirmed, unspentOutput.satoshis);
            }
            else if (utxoInUse) {
                let unspentOutput = new unspent_output_1.UnspentOutput({
                    txid: utxo.transactionId,
                    address,
                    vout: utxo.index,
                    scriptPubKey: utxo.scriptPublicKey.scriptPublicKey,
                    scriptPublicKeyVersion: utxo.scriptPublicKey.version,
                    satoshis: +utxo.amount,
                    blockDaaScore: utxo.blockDaaScore,
                    isCoinbase: utxo.isCoinbase
                });
                this.utxos.used.set(utxoId, unspentOutput);
            }
        });
        if (utxoIds.length) {
            this.logger.utxodebug(`adding ${utxoIds.length} UTXO entries:\n`, utxoIds);
            this.logger.utxo(`incoming ${utxoIds.length} UTXO entries`);
        }
        this.wallet.txStore.addAddressUTXOs(address, utxos);
        return utxoIds;
    }
    get logger() {
        return this.wallet.logger;
    }
    remove(utxoIds) {
        this.release(utxoIds);
        let { blueScore } = this.wallet;
        let utxo;
        utxoIds.forEach(id => {
            utxo = this.utxos.confirmed.get(id);
            if (utxo) {
                this.utxos.confirmed.delete(id);
                this.wallet.adjustBalance(true, -utxo.satoshis);
            }
            utxo = this.utxos.pending.get(id);
            if (utxo) {
                this.utxos.pending.delete(id);
                this.wallet.adjustBalance(false, -utxo.satoshis);
                //duplicate tx issue handling
                if (utxo.blockDaaScore - blueScore < 70) {
                    let apiUTXO = {
                        transactionId: utxo.txId,
                        amount: utxo.satoshis,
                        scriptPublicKey: {
                            version: utxo.scriptPublicKeyVersion,
                            scriptPublicKey: utxo.scriptPubKey
                        },
                        blockDaaScore: utxo.blockDaaScore,
                        index: utxo.outputIndex,
                        isCoinbase: utxo.isCoinbase
                    };
                    this.wallet.txStore.removePendingUTXO(apiUTXO, utxo.address.toString());
                }
            }
        });
    }
    clearUsed() {
        this.inUse = [];
        this.utxos.used.clear();
        this.wallet.updateDebugInfo();
        this.wallet.emitCache();
    }
    clearMissing() {
        const { confirmed, pending, used } = this.utxos;
        let missing = this.inUse.filter(utxoId => {
            return !(confirmed.has(utxoId) || pending.has(utxoId) || used.has(utxoId));
        });
        if (!missing.length)
            return false;
        this.release(missing);
        return true;
    }
    release(utxoIdsToEnable) {
        // assigns new array without any utxoIdsToEnable
        this.inUse = this.inUse.filter((utxoId) => !utxoIdsToEnable.includes(utxoId));
        utxoIdsToEnable.forEach(utxoId => {
            this.utxos.used.delete(utxoId);
        });
        this.wallet.updateDebugInfo();
        this.wallet.emitCache();
        // this.updateUtxoBalance();
    }
    updateUtxoBalance() {
        const { blueScore } = this.wallet;
        [...this.utxos.pending.values()].forEach(utxo => {
            if (blueScore - utxo.blockDaaScore < (utxo.isCoinbase ? exports.COINBASE_CFM_COUNT : exports.CONFIRMATION_COUNT))
                return;
            this.utxos.pending.delete(utxo.txId + utxo.outputIndex);
            this.wallet.adjustBalance(false, -utxo.satoshis, false);
            this.utxos.confirmed.set(utxo.txId + utxo.outputIndex, utxo);
            this.wallet.adjustBalance(true, utxo.satoshis);
        });
    }
    clear() {
        this.utxos.confirmed.clear();
        this.utxos.pending.clear();
        this.utxos.used.clear();
        this.inUse = [];
        this.availableBalance = 0;
        this.utxoStorage = {};
        this.logger.info('UTXO set cleared.');
    }
    updateUsed(utxos) {
        utxos.forEach(utxo => {
            this.inUse.push(utxo.id);
            this.utxos.used.set(utxo.txId, utxo);
        });
        this.wallet.updateDebugInfo();
        this.wallet.emitCache();
    }
    /**
     * Naively select UTXOs.
     * @param txAmount Provide the amount that the UTXOs should cover.
     * @throws Error message if the UTXOs can't cover the `txAmount`
     */
    selectUtxos(txAmount) {
        const utxos = [];
        const utxoIds = [];
        let totalVal = 0;
        let list = [...this.utxos.confirmed.values()];
        list = list.filter((utxo) => {
            return !this.inUse.includes(utxo.id);
        });
        list.sort((a, b) => {
            return a.blockDaaScore - b.blockDaaScore || b.satoshis - a.satoshis || a.txId.localeCompare(b.txId) || a.outputIndex - b.outputIndex;
        });
        let mass = 0;
        for (const utxo of list) {
            //console.log("info",`UTXO ID: ${utxoId}  , UTXO: ${utxo}`);
            //if (!this.inUse.includes(utxoId)) {
            utxoIds.push(utxo.id);
            utxos.push(utxo);
            mass += utxo.mass;
            totalVal += utxo.satoshis;
            //}
            if (totalVal >= txAmount)
                break;
        }
        if (totalVal < txAmount)
            throw new Error(`Insufficient balance - need: ${KLS(txAmount)} KLS, available: ${KLS(totalVal)} KLS`);
        return {
            utxoIds,
            utxos,
            mass
        };
    }
    /**
     * Naively collect UTXOs.
     * @param maxCount Provide the max UTXOs count.
     */
    collectUtxos(maxCount = 10000) {
        const utxos = [];
        const utxoIds = [];
        let totalVal = 0;
        let list = [...this.utxos.confirmed.values()];
        list = list.filter((utxo) => {
            return !this.inUse.includes(utxo.id);
        });
        list.sort((a, b) => {
            return a.blockDaaScore - b.blockDaaScore || b.satoshis - a.satoshis || a.txId.localeCompare(b.txId) || a.outputIndex - b.outputIndex;
        });
        let maxMass = wallet_1.Wallet.MaxMassUTXOs;
        let mass = 0;
        for (const utxo of list) {
            if (utxos.length >= maxCount || mass + utxo.mass >= maxMass)
                break;
            utxoIds.push(utxo.id);
            utxos.push(utxo);
            totalVal += utxo.satoshis;
            mass += utxo.mass;
        }
        //console.log("maxMass:"+maxMass, "mass:"+mass)
        return {
            utxoIds,
            utxos,
            amount: totalVal,
            mass
        };
    }
    syncAddressesUtxos(addresses) {
        return __awaiter(this, void 0, void 0, function* () {
            const newAddresses = addresses.map(address => {
                if (this.addressesUtxoSyncStatuses.has(address))
                    return;
                this.addressesUtxoSyncStatuses.set(address, false);
                return address;
            }).filter(address => address);
            //in sync process addressDiscovery calls findUtxos
            if (!newAddresses.length || (this.wallet.syncInProggress && !this.wallet.options.disableAddressDerivation))
                return;
            yield this.wallet.findUtxos(newAddresses);
            if (!this.wallet.syncOnce)
                yield this.utxoSubscribe();
        });
    }
    utxoSubscribe() {
        return __awaiter(this, void 0, void 0, function* () {
            let addresses = [];
            this.addressesUtxoSyncStatuses.forEach((sent, address) => {
                //if(sent)
                //  return
                //  !!!FIXME prevent multiple address subscriptions
                //if(!this.addressesUtxoSyncStatuses.get(address)) {
                //this.addressesUtxoSyncStatuses.set(address, true);
                addresses.push(address);
                //}
            });
            if (!addresses.length)
                return addresses;
            //console.log(`[${this.wallet.network}] !!! +++++++++++++++ SUBSCRIBING TO ADDRESSES :)\n`,addresses);
            let utxoChangedRes = yield this.wallet.api.subscribeUtxosChanged(addresses, this.onUtxosChanged.bind(this))
                .catch((error) => {
                console.log(`[${this.wallet.network}] RPC ERROR in uxtoSync! while registering addresses:`, error, addresses);
                addresses.map(address => {
                    this.addressesUtxoSyncStatuses.set(address, false);
                });
            });
            //console.log("utxoSync:utxoChangedRes:", utxoChangedRes, "\n utxoSync addresses:", addresses)
            return addresses;
        });
    }
    onUtxosChanged(added, removed) {
        // console.log("onUtxosChanged:res", added, removed)
        added.forEach((utxos, address) => {
            //this.logger.log('info', `${address}: ${utxos.length} utxos found.+=+=+=+=+=+=+++++=======+===+====+====+====+`);
            if (!utxos.length)
                return;
            if (!this.utxoStorage[address]) {
                this.utxoStorage[address] = utxos;
            }
            else {
                let txid2Utxo = {};
                utxos.forEach(utxo => {
                    txid2Utxo[utxo.transactionId + utxo.index] = utxo;
                });
                let oldUtxos = this.utxoStorage[address].filter(utxo => {
                    return !txid2Utxo[utxo.transactionId + utxo.index];
                });
                this.utxoStorage[address] = [...oldUtxos, ...utxos];
            }
            this.add(utxos, address);
        });
        this.wallet.txStore.addFromUTXOs(added);
        let utxoIds = [];
        removed.forEach((utxos, address) => {
            let txid2Outpoint = {};
            utxos.forEach(utxo => {
                txid2Outpoint[utxo.transactionId + utxo.index] = utxo;
                utxoIds.push(utxo.transactionId + utxo.index);
            });
            if (!this.utxoStorage[address])
                return;
            this.utxoStorage[address] = this.utxoStorage[address].filter(utxo => {
                return !txid2Outpoint[utxo.transactionId + utxo.index];
            });
        });
        if (utxoIds.length)
            this.remove(utxoIds);
        const isActivityOnReceiveAddr = this.utxoStorage[this.wallet.receiveAddress] !== undefined;
        if (isActivityOnReceiveAddr)
            this.wallet.addressManager.receiveAddress.next();
        //this.updateUtxoBalance();
        this.wallet.emit("utxo-change", { added, removed });
    }
    isOur(utxo) {
        return (!!this.wallet.transactions[utxo.txId]) || this.isOurChange(utxo);
    }
    isOurChange(utxo) {
        return this.wallet.addressManager.isOurChange(String(utxo.address));
    }
    get count() {
        return this.utxos.confirmed.size + this.utxos.pending.size;
    }
    get confirmedCount() {
        return this.utxos.confirmed.size;
    }
}
exports.UtxoSet = UtxoSet;
