// @pugdag/wallet/dist/wallet/wallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallet = exports.COINBASE_CFM_COUNT = exports.CONFIRMATION_COUNT = exports.COMPOUND_UTXO_MAX_COUNT = exports.pugdagcore = void 0;
const Mnemonic = require('bitcore-mnemonic');
const pugdagcore = __importStar(require("../../../core-lib"));
exports.pugdagcore = pugdagcore;
const helper = __importStar(require("../utils/helper"));
__exportStar(require("./storage"), exports);
__exportStar(require("./error"), exports);
const crypto_1 = require("./crypto");
const PUG = helper.PUG;
const logger_1 = require("../utils/logger");
const address_manager_1 = require("./address-manager");
const utxo_1 = require("./utxo");
Object.defineProperty(exports, "CONFIRMATION_COUNT", { enumerable: true, get: function () { return utxo_1.CONFIRMATION_COUNT; } });
Object.defineProperty(exports, "COINBASE_CFM_COUNT", { enumerable: true, get: function () { return utxo_1.COINBASE_CFM_COUNT; } });
const tx_store_1 = require("./tx-store");
const cache_store_1 = require("./cache-store");
const api_1 = require("./api");
const config_json_1 = require("../config.json");
const event_target_impl_1 = require("./event-target-impl");
const BALANCE_CONFIRMED = Symbol();
const BALANCE_PENDING = Symbol();
const BALANCE_TOTAL = Symbol();
const COMPOUND_UTXO_MAX_COUNT = 500;
exports.COMPOUND_UTXO_MAX_COUNT = COMPOUND_UTXO_MAX_COUNT;
const SompiPerPugdag = 100000000;
// MaxSompi is the maximum transaction amount allowed in sompi.
const MaxSompi = 21000000 * SompiPerPugdag;
/** Class representing an HDWallet with derivable child addresses */
class Wallet extends event_target_impl_1.EventTargetImpl {
    static PUG(v) {
        return PUG(v);
    }
    static initRuntime() {
        return pugdagcore.initRuntime();
    }
    /**
     * Converts a mnemonic to a new wallet.
     * @param seedPhrase The 12 word seed phrase.
     * @returns new Wallet
     */
    static fromMnemonic(seedPhrase, networkOptions, options = {}) {
        if (!networkOptions || !networkOptions.network)
            throw new Error(`fromMnemonic(seedPhrase,networkOptions): missing network argument`);
        const privKey = new Mnemonic(seedPhrase.trim()).toHDPrivateKey().toString();
        const wallet = new this(privKey, seedPhrase, networkOptions, options);
        return wallet;
    }
    /**
     * Creates a new Wallet from encrypted wallet data.
     * @param password the password the user encrypted their seed phrase with
     * @param encryptedMnemonic the encrypted seed phrase from local storage
     * @throws Will throw "Incorrect password" if password is wrong
     */
    static import(password, encryptedMnemonic, networkOptions, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const decrypted = yield crypto_1.Crypto.decrypt(password, encryptedMnemonic);
            const savedWallet = JSON.parse(decrypted);
            const myWallet = new this(savedWallet.privKey, savedWallet.seedPhrase, networkOptions, options);
            return myWallet;
        });
    }
    get balance() {
        return {
            available: this[BALANCE_CONFIRMED],
            pending: this[BALANCE_PENDING],
            total: this[BALANCE_CONFIRMED] + this[BALANCE_PENDING]
        };
    }
    /**
     * Set by addressManager
     */
    get receiveAddress() {
        return this.addressManager.receiveAddress.current.address;
    }
    get changeAddress() {
        return this.addressManager.changeAddress.current.address;
    }
    /** Create a wallet.
     * @param walletSave (optional)
     * @param walletSave.privKey Saved wallet's private key.
     * @param walletSave.seedPhrase Saved wallet's seed phrase.
     */
    constructor(privKey, seedPhrase, networkOptions, options = {}) {
        super();
        this.disableBalanceNotifications = false;
        /**
         * Current network.
         */
        this.network = 'pugdag';
        /**
         * Default fee
         */
        this.defaultFee = 1; //per byte
        this.subnetworkId = "0000000000000000000000000000000000000000"; //hex string
        this.last_tx_ = '';
        /**
         * Current API endpoint for selected network
         */
        this.apiEndpoint = 'localhost:33455';
        this.blueScore = -1;
        this.syncVirtualSelectedParentBlueScoreStarted = false;
        this.syncInProggress = false;
        /* eslint-disable */
        this.pendingInfo = {
            transactions: {},
            get amount() {
                const transactions = Object.values(this.transactions);
                if (transactions.length === 0)
                    return 0;
                return transactions.reduce((prev, cur) => prev + cur.amount + cur.fee, 0);
            },
            add(id, tx) {
                this.transactions[id] = tx;
            }
        };
        /**
         * Transactions sorted by hash.
         */
        this.transactions = {};
        /**
         * Transaction arrays keyed by address.
         */
        this.transactionsStorage = {};
        this[_a] = 0;
        this[_b] = 0;
        this[_c] = 0;
        /**
         * Emit wallet balance.
         */
        this.lastBalanceNotification = { available: 0, pending: 0 };
        this.debugInfo = { inUseUTXOs: { satoshis: 0, count: 0 } };
        this.lastAddressNotification = {};
        //UTXOsPollingStarted:boolean = false;
        this.emitedUTXOs = new Set();
        this.loggerLevel = 0;
        this.logger = (0, logger_1.CreateLogger)('PugdagWallet');
        this.api = new api_1.PugdagAPI();
        //@ts-ignore
        //postMessage({error:new ApiError("test") })
        let defaultOpt = {
            skipSyncBalance: false,
            syncOnce: false,
            addressDiscoveryExtent: 150,
            logLevel: 'info',
            disableAddressDerivation: false,
            checkGRPCFlags: false,
            minimumRelayTransactionFee: 1000,
            updateTxTimes: true
        };
        // console.log("CREATING WALLET FOR NETWORK", this.network);
        this.options = Object.assign(Object.assign({}, defaultOpt), options);
        //this.options.addressDiscoveryExtent = 500;
        this.setLogLevel(this.options.logLevel);
        this.network = networkOptions.network;
        this.defaultFee = networkOptions.defaultFee || this.defaultFee;
        if (networkOptions.rpc)
            this.api.setRPC(networkOptions.rpc);
        if (privKey && seedPhrase) {
            this.HDWallet = new pugdagcore.HDPrivateKey(privKey);
            this.mnemonic = seedPhrase;
        }
        else {
            const temp = new Mnemonic(Mnemonic.Words.ENGLISH);
            this.mnemonic = temp.toString();
            this.HDWallet = new pugdagcore.HDPrivateKey(temp.toHDPrivateKey().toString());
        }
        this.uid = this.createUID();
        this.utxoSet = new utxo_1.UtxoSet(this);
        this.txStore = new tx_store_1.TXStore(this);
        this.cacheStore = new cache_store_1.CacheStore(this);
        //this.utxoSet.on("balance-update", this.updateBalance.bind(this));
        this.addressManager = new address_manager_1.AddressManager(this.HDWallet, this.network);
        if (this.options.disableAddressDerivation)
            this.addressManager.receiveAddress.next();
        //this.initAddressManager();
        //this.sync(this.options.syncOnce);
        this.connectSignal = helper.Deferred();
        this.api.on("connect", () => {
            this.onApiConnect();
        });
        this.api.on("disconnect", () => {
            this.onApiDisconnect();
        });
    }
    createUID() {
        const { privateKey } = this.HDWallet.deriveChild(`m/44'/972/0'/1'/0'`);
        let address = privateKey.toAddress(this.network).toString().split(":")[1];
        return helper.createHash(address);
    }
    onApiConnect() {
        return __awaiter(this, void 0, void 0, function* () {
            this.connectSignal.resolve();
            let { connected } = this;
            this.connected = true;
            this.logger.info("gRPC connected");
            this.emit("api-connect");
            if (this.syncSignal && connected !== undefined) { //if sync was called
                this.logger.info("starting wallet re-sync ...");
                yield this.sync(this.syncOnce);
            }
        });
    }
    onApiDisconnect() {
        this.connected = false;
        this.syncVirtualSelectedParentBlueScoreStarted = false;
        this.logger.verbose("gRPC disconnected");
        this.emit("api-disconnect");
    }
    update(syncOnce = true) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.sync(syncOnce);
        });
    }
    waitOrSync() {
        if (this.syncSignal)
            return this.syncSignal;
        return this.sync();
    }
    sync(syncOnce = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            this.syncSignal = helper.Deferred();
            yield this.connectSignal;
            if (syncOnce === undefined)
                syncOnce = this.options.syncOnce;
            syncOnce = !!syncOnce;
            this.syncInProggress = true;
            this.emit("sync-start");
            yield this.txStore.restore();
            yield this.cacheStore.restore();
            const ts0 = Date.now();
            this.logger.info(`sync ... starting wallet sync`); // ${syncOnce?'(monitoring disabled)':''}`);
            //this.logger.info(`sync ............ started, syncOnce:${syncOnce}`)
            //if last time syncOnce was OFF we have subscriptions to utxo-change
            if (this.syncOnce === false && syncOnce) {
                throw new Error("Wallet sync process already running.");
            }
            this.syncOnce = syncOnce;
            this.initAddressManager();
            yield this.initBlueScoreSync(syncOnce)
                .catch(e => {
                this.logger.info("syncVirtualSelectedParentBlueScore:error", e);
            });
            if (this.options.disableAddressDerivation) {
                this.logger.warn('sync ... running with address discovery disabled');
                this.utxoSet.syncAddressesUtxos([this.receiveAddress]);
            }
            else {
                yield this.addressDiscovery(this.options.addressDiscoveryExtent)
                    .catch(e => {
                    this.logger.info("addressDiscovery:error", e);
                });
            }
            this.syncInProggress = false;
            if (!syncOnce)
                yield this.utxoSet.utxoSubscribe();
            const ts1 = Date.now();
            const delta = ((ts1 - ts0) / 1000).toFixed(1);
            this.logger.info(`sync ... ${this.utxoSet.count} UTXO entries found`);
            this.logger.info(`sync ... indexed ${this.addressManager.receiveAddress.counter} receive and ${this.addressManager.changeAddress.counter} change addresses`);
            this.logger.info(`sync ... finished (sync done in ${delta} seconds)`);
            this.emit("sync-finish");
            const { available, pending, total } = this.balance;
            this.emit("ready", {
                available, pending, total,
                confirmedUtxosCount: this.utxoSet.confirmedCount
            });
            this.emitBalance();
            this.emitAddress();
            this.txStore.emitTxs();
            this.syncSignal.resolve();
            if (!this.utxoSet.clearMissing())
                this.updateDebugInfo();
        });
    }
    getVirtualSelectedParentBlueScore() {
        return this.api.getVirtualSelectedParentBlueScore();
    }
    getVirtualDaaScore() {
        return this.api.getVirtualDaaScore();
    }
    initBlueScoreSync(once = false) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.syncVirtualSelectedParentBlueScoreStarted)
                return;
            this.syncVirtualSelectedParentBlueScoreStarted = true;
            let r = yield this.getVirtualDaaScore();
            let { virtualDaaScore: blueScore } = r;
            console.log("getVirtualSelectedParentBlueScore :result", r);
            this.blueScore = blueScore;
            this.emit("blue-score-changed", { blueScore });
            this.utxoSet.updateUtxoBalance();
            if (once) {
                this.syncVirtualSelectedParentBlueScoreStarted = false;
                return;
            }
            this.api.subscribeVirtualDaaScoreChanged((result) => {
                let { virtualDaaScore } = result;
                //console.log("subscribeVirtualSelectedParentBlueScoreChanged:result", result)
                this.blueScore = virtualDaaScore;
                this.emit("blue-score-changed", {
                    blueScore: virtualDaaScore
                });
                this.utxoSet.updateUtxoBalance();
            }).then(r => {
                console.log("subscribeVirtualDaaScoreChanged:responce", r);
            }, e => {
                console.log("subscribeVirtualDaaScoreChanged:error", e);
            });
        });
    }
    initAddressManager() {
        if (this.addressManagerInitialized)
            return;
        this.addressManagerInitialized = true;
        this.addressManager.on("new-address", detail => {
            if (!this.syncInProggress) {
                this.emitAddress();
            }
            //console.log("new-address", detail)
            if (this.options.skipSyncBalance)
                return;
            //console.log("new-address:detail", detail)
            const { address, type } = detail;
            this.utxoSet.syncAddressesUtxos([address]);
        });
        if (!this.receiveAddress) {
            this.addressManager.receiveAddress.next();
        }
    }
    startUpdatingTransactions(version = undefined) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.txStore.startUpdatingTransactions(version);
        });
    }
    /**
     * Set rpc provider
     * @param rpc
     */
    setRPC(rpc) {
        this.api.setRPC(rpc);
    }
    /*
    setStorageType(type:StorageType){
        this.storage.setType(type);
    }
    setStorageFolder(folder:string){
        this.storage.setFolder(folder);
    }
    setStorageFileName(fileName:string){
        this.storage.setFileName(fileName);
    }
    */
    /*
    _storage: typeof storageClasses.Storage|undefined;

    setStoragePassword(password: string) {
        if (!this.storage)
            throw new Error("Please init storage")
        this.storage.setPassword(password);
    }
    get storage(): typeof storageClasses.Storage | undefined {
        return this._storage;
    }

    openFileStorage(fileName: string, password: string, folder: string = '') {
        let storage = CreateStorage();
        if (folder)
            storage.setFolder(folder);
        storage.setFileName(fileName);
        storage.setPassword(password);
        this._storage = storage;
    }
    */
    /**
     * Queries API for address[] UTXOs. Adds tx to transactions storage. Also sorts the entire transaction set.
     * @param addresses
     */
    findUtxos(addresses, debug = false) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.verbose(`scanning UTXO entries for ${addresses.length} addresses`);
            const utxosMap = yield this.api.getUtxosByAddresses(addresses);
            const addressesWithUTXOs = [];
            const txID2Info = new Map();
            if (debug) {
                utxosMap.forEach((utxos, address) => {
                    // utxos.sort((b, a)=> a.index-b.index)
                    utxos.map(t => {
                        let info = txID2Info.get(t.transactionId);
                        if (!info) {
                            info = {
                                utxos: [],
                                address
                            };
                            txID2Info.set(t.transactionId, info);
                        }
                        info.utxos.push(t);
                    });
                });
            }
            utxosMap.forEach((utxos, address) => {
                // utxos.sort((b, a)=> a.index-b.index)
                this.logger.verbose(`${address} - ${utxos.length} UTXO entries found`);
                if (utxos.length !== 0) {
                    this.disableBalanceNotifications = true;
                    this.utxoSet.utxoStorage[address] = utxos;
                    this.utxoSet.add(utxos, address);
                    addressesWithUTXOs.push(address);
                    this.disableBalanceNotifications = false;
                    this.emitBalance();
                }
            });
            const isActivityOnReceiveAddr = this.utxoSet.utxoStorage[this.receiveAddress] !== undefined;
            if (isActivityOnReceiveAddr) {
                this.addressManager.receiveAddress.next();
            }
            return {
                addressesWithUTXOs,
                txID2Info
            };
        });
    }
    adjustBalance(isConfirmed, amount, notify = true) {
        const { available, pending } = this.balance;
        if (isConfirmed) {
            this[BALANCE_CONFIRMED] += amount;
        }
        else {
            this[BALANCE_PENDING] += amount;
        }
        this[BALANCE_TOTAL] = this[BALANCE_CONFIRMED] + this[BALANCE_PENDING];
        if (notify === false)
            return;
        const { available: _available, pending: _pending } = this.balance;
        if (!this.syncInProggress && !this.disableBalanceNotifications && (available != _available || pending != _pending))
            this.emitBalance();
    }
    emitBalance() {
        const { available, pending, total } = this.balance;
        const { available: _available, pending: _pending } = this.lastBalanceNotification;
        if (available == _available && pending == _pending)
            return;
        this.lastBalanceNotification = { available, pending };
        this.logger.debug(`balance available: ${available} pending: ${pending}`);
        this.emit("balance-update", {
            available,
            pending,
            total,
            confirmedUtxosCount: this.utxoSet.confirmedCount
        });
    }
    updateDebugInfo() {
        let inUseUTXOs = { satoshis: 0, count: 0 };
        let { confirmed, pending, used } = this.utxoSet.utxos || {};
        this.utxoSet.inUse.map(utxoId => {
            var _d, _e, _f;
            inUseUTXOs.satoshis += ((_d = confirmed.get(utxoId)) === null || _d === void 0 ? void 0 : _d.satoshis) ||
                ((_e = pending.get(utxoId)) === null || _e === void 0 ? void 0 : _e.satoshis) ||
                ((_f = used.get(utxoId)) === null || _f === void 0 ? void 0 : _f.satoshis) || 0;
        });
        inUseUTXOs.count = this.utxoSet.inUse.length;
        this.debugInfo = { inUseUTXOs };
        this.emit("debug-info", { debugInfo: this.debugInfo });
    }
    clearUsedUTXOs() {
        this.utxoSet.clearUsed();
    }
    emitCache() {
        let { cache } = this;
        this.emit("state-update", { cache });
    }
    emitAddress() {
        const receive = this.receiveAddress;
        const change = this.changeAddress;
        let { receive: _receive, change: _change } = this.lastAddressNotification;
        if (receive == _receive && change == _change)
            return;
        this.lastAddressNotification = { receive, change };
        this.emit("new-address", {
            receive, change
        });
    }
    /**
     * Updates the selected network
     * @param network name of the network
     */
    updateNetwork(network) {
        return __awaiter(this, void 0, void 0, function* () {
            this.demolishWalletState(network.prefix);
            this.network = network.prefix;
            this.apiEndpoint = network.apiBaseUrl;
        });
    }
    demolishWalletState(networkPrefix = this.network) {
        this.utxoSet.clear();
        this.addressManager = new address_manager_1.AddressManager(this.HDWallet, networkPrefix);
        //this.pendingInfo.transactions = {};
        this.transactions = {};
        this.transactionsStorage = {};
    }
    scanMoreAddresses(count = 100, debug = false, receiveStart = -1, changeStart = -1) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.syncInProggress)
                return { error: "Sync in progress", code: "SYNC-IN-PROGRESS" };
            if (receiveStart < 0)
                receiveStart = this.addressManager.receiveAddress.counter;
            if (changeStart < 0)
                changeStart = this.addressManager.changeAddress.counter;
            this.syncInProggress = true;
            this.emit("scan-more-addresses-started", { receiveStart, changeStart });
            let error = false;
            let res = yield this.addressDiscovery(this.options.addressDiscoveryExtent, debug, receiveStart, changeStart, count)
                .catch(e => {
                this.logger.info("addressDiscovery:error", e);
                error = e;
            });
            this.syncInProggress = false;
            if (!this.syncOnce)
                this.utxoSet.utxoSubscribe();
            this.emit("scan-more-addresses-ended", { error });
            if (error)
                return { error, code: "ADDRESS-DISCOVERY" };
            let { highestIndex = null, endPoints = null } = res || {};
            this.logger.info("scanMoreAddresses:highestIndex", highestIndex);
            this.logger.info("scanMoreAddresses:endPoints", endPoints);
            this.emit("scan-more-addresses-ended", {
                receiveFinal: this.addressManager.receiveAddress.counter - 1,
                changeFinal: this.addressManager.changeAddress.counter - 1
            });
            return {
                code: "SUCCESS",
                receive: {
                    start: receiveStart,
                    end: (endPoints === null || endPoints === void 0 ? void 0 : endPoints.receive) || receiveStart + count,
                    final: this.addressManager.receiveAddress.counter - 1
                },
                change: {
                    start: changeStart,
                    end: (endPoints === null || endPoints === void 0 ? void 0 : endPoints.change) || changeStart + count,
                    final: this.addressManager.changeAddress.counter - 1
                }
            };
        });
    }
    /**
     * Derives receiveAddresses and changeAddresses and checks their transactions and UTXOs.
     * @param threshold stop discovering after `threshold` addresses with no activity
     */
    addressDiscovery(threshold = 64, debug = false, receiveStart = 0, changeStart = 0, count = 0) {
        var _d;
        return __awaiter(this, void 0, void 0, function* () {
            let addressList = [];
            let debugInfo = null;
            this.logger.info(`sync ... running address discovery, threshold:${threshold}`);
            const cacheIndexes = (_d = this.cacheStore.getAddressIndexes()) !== null && _d !== void 0 ? _d : { receive: 0, change: 0 };
            this.logger.info(`sync ...cacheIndexes: receive:${cacheIndexes.receive}, change:${cacheIndexes.change}`);
            let highestIndex = {
                receive: this.addressManager.receiveAddress.counter - 1,
                change: this.addressManager.changeAddress.counter - 1
            };
            let endPoints = {
                receive: 0,
                change: 0
            };
            let maxOffset = {
                receive: receiveStart + count,
                change: changeStart + count
            };
            const doDiscovery = (n, deriveType, offset) => __awaiter(this, void 0, void 0, function* () {
                this.logger.info(`sync ... scanning ${offset} - ${offset + n} ${deriveType} addresses`);
                this.emit("sync-progress", {
                    start: offset,
                    end: offset + n,
                    addressType: deriveType
                });
                const derivedAddresses = this.addressManager.getAddresses(n, deriveType, offset);
                const addresses = derivedAddresses.map((obj) => obj.address);
                addressList = [...addressList, ...addresses];
                this.logger.verbose(`${deriveType}: address data for derived indices ${derivedAddresses[0].index}..${derivedAddresses[derivedAddresses.length - 1].index}`);
                // if (this.loggerLevel > 0)
                // 	this.logger.verbose("addressDiscovery: findUtxos for addresses::", addresses)
                const { addressesWithUTXOs, txID2Info } = yield this.findUtxos(addresses, debug);
                if (!debugInfo)
                    debugInfo = txID2Info;
                if (addressesWithUTXOs.length === 0) {
                    // address discovery complete
                    const lastAddressIndexWithTx = highestIndex[deriveType]; //offset - (threshold - n) - 1;
                    this.logger.verbose(`${deriveType}: address discovery complete`);
                    this.logger.verbose(`${deriveType}: last activity on address #${lastAddressIndexWithTx}`);
                    this.logger.verbose(`${deriveType}: no activity from ${offset}..${offset + n}`);
                    if (offset >= maxOffset[deriveType] && offset >= cacheIndexes[deriveType]) {
                        endPoints[deriveType] = offset + n;
                        return lastAddressIndexWithTx;
                    }
                }
                // else keep doing discovery
                const index = derivedAddresses
                    .filter((obj) => addressesWithUTXOs.includes(obj.address))
                    .reduce((prev, cur) => Math.max(prev, cur.index), highestIndex[deriveType]);
                highestIndex[deriveType] = index;
                return doDiscovery(n, deriveType, offset + n);
            });
            const highestReceiveIndex = yield doDiscovery(threshold, 'receive', receiveStart);
            const highestChangeIndex = yield doDiscovery(threshold, 'change', changeStart);
            this.addressManager.receiveAddress.advance(highestReceiveIndex + 1);
            this.addressManager.changeAddress.advance(highestChangeIndex + 1);
            this.logger.verbose(`receive address index: ${highestReceiveIndex}; change address index: ${highestChangeIndex}`, `receive-address-index: ${this.addressManager.receiveAddress.counter}; change address index: ${this.addressManager.changeAddress.counter}`);
            if (!this.syncOnce && !this.syncInProggress)
                yield this.utxoSet.utxoSubscribe();
            this.runStateChangeHooks();
            let addressIndexes = {
                receive: Math.max(cacheIndexes.receive, this.addressManager.receiveAddress.counter),
                change: Math.max(cacheIndexes.change, this.addressManager.changeAddress.counter)
            };
            this.logger.info(`sync ...new cache: receive:${addressIndexes.receive}, change:${addressIndexes.change}`);
            this.cacheStore.setAddressIndexes(addressIndexes);
            this.emit("sync-end", addressIndexes);
            return { highestIndex, endPoints, debugInfo };
        });
    }
    // TODO: convert amount to sompis aka satoshis
    // TODO: bn
    /**
     * Compose a serialized, signed transaction
     * @param obj
     * @param obj.toAddr To address in cashaddr format (e.g. pugdagtest:qq0d6h0prjm5mpdld5pncst3adu0yam6xch4tr69k2)
     * @param obj.amount Amount to send in sompis (100000000 (1e8) sompis in 1 PUG)
     * @param obj.fee Fee for miners in sompis
     * @param obj.changeAddrOverride Use this to override automatic change address derivation
     * @throws if amount is above `Number.MAX_SAFE_INTEGER`
     */
    composeTx({ toAddr, amount, fee = config_json_1.DEFAULT_FEE, changeAddrOverride, skipSign = false, privKeysInfo = false, compoundingUTXO = false, compoundingUTXOMaxCount = COMPOUND_UTXO_MAX_COUNT }) {
        // TODO: bn!
        amount = parseInt(amount);
        fee = parseInt(fee);
        // if (this.loggerLevel > 0) {
        // 	for (let i = 0; i < 100; i++)
        // 		console.log('Wallet transaction request for', amount, typeof amount);
        // }
        //if (!Number.isSafeInteger(amount)) throw new Error(`Amount ${amount} is too large`);
        let utxos, utxoIds, mass;
        if (compoundingUTXO) {
            ({ utxos, utxoIds, amount, mass } = this.utxoSet.collectUtxos(compoundingUTXOMaxCount));
        }
        else {
            ({ utxos, utxoIds, mass } = this.utxoSet.selectUtxos(amount + fee));
        }
        //if(mass > Wallet.MaxMassUTXOs){
        //	throw new Error(`Maximum number of inputs (UTXOs) reached. Please reduce this transaction amount.`);
        //}
        const privKeys = utxos.reduce((prev, cur) => {
            return [this.addressManager.all[String(cur.address)], ...prev];
        }, []);
        this.logger.info("utxos.length", utxos.length);
        const changeAddr = changeAddrOverride || this.addressManager.changeAddress.next();
        try {
            const tx = new pugdagcore.Transaction()
                .from(utxos)
                .to(toAddr, amount)
                .setVersion(0)
                .fee(fee)
                .change(changeAddr);
            if (!skipSign)
                tx.sign(privKeys, pugdagcore.crypto.Signature.SIGHASH_ALL, 'schnorr');
            //window.txxxx = tx;
            return {
                tx: tx,
                id: tx.id,
                rawTx: tx.toString(),
                utxoIds,
                amount,
                fee,
                utxos,
                toAddr,
                privKeys: privKeysInfo ? privKeys : []
            };
        }
        catch (e) {
            console.log("composeTx:error", e);
            // !!! FIXME
            if (!changeAddrOverride)
                this.addressManager.changeAddress.reverse();
            throw e;
        }
    }
    minimumRequiredTransactionRelayFee(mass) {
        let minimumFee = (mass * this.options.minimumRelayTransactionFee) / 1000;
        if (minimumFee == 0 && this.options.minimumRelayTransactionFee > 0) {
            minimumFee = this.options.minimumRelayTransactionFee;
        }
        // Set the minimum fee to the maximum possible value if the calculated
        // fee is not in the valid range for monetary amounts.
        if (minimumFee > MaxSompi) {
            minimumFee = MaxSompi;
        }
        return minimumFee;
    }
    /*
    validateAddress(addr:string):boolean{
        let address = new pugdagcore.Address(addr);
        return address.type == "pubkey";
    }
    */
    /**
     * Estimate transaction fee. Returns transaction data.
     * @param txParams
     * @param txParams.toAddr To address in cashaddr format (e.g. pugdagtest:qq0d6h0prjm5mpdld5pncst3adu0yam6xch4tr69k2)
     * @param txParams.amount Amount to send in sompis (100000000 (1e8) sompis in 1 PUG)
     * @param txParams.fee Fee for miners in sompis
     * @throws `FetchError` if endpoint is down. API error message if tx error. Error if amount is too large to be represented as a javascript number.
     */
    estimateTransaction(txParamsArg) {
        return __awaiter(this, void 0, void 0, function* () {
            let address = this.addressManager.changeAddress.current.address;
            if (!address) {
                address = this.addressManager.changeAddress.next();
            }
            txParamsArg.changeAddrOverride = address;
            return this.composeTxAndNetworkFeeInfo(txParamsArg);
        });
    }
    composeTxAndNetworkFeeInfo(txParamsArg) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.waitOrSync();
            if (!txParamsArg.fee)
                txParamsArg.fee = 0;
            this.logger.info(`tx ... sending to ${txParamsArg.toAddr}`);
            this.logger.info(`tx ... amount: ${PUG(txParamsArg.amount)} user fee: ${PUG(txParamsArg.fee)} max data fee: ${PUG(txParamsArg.networkFeeMax || 0)}`);
            //if(!this.validateAddress(txParamsArg.toAddr)){
            //	throw new Error("Invalid address")
            //}
            let txParams = Object.assign({}, txParamsArg);
            const networkFeeMax = txParams.networkFeeMax || 0;
            let calculateNetworkFee = !!txParams.calculateNetworkFee;
            let inclusiveFee = !!txParams.inclusiveFee;
            const { skipSign = true, privKeysInfo = false } = txParams;
            txParams.skipSign = skipSign;
            txParams.privKeysInfo = privKeysInfo;
            //console.log("calculateNetworkFee:", calculateNetworkFee, "inclusiveFee:", inclusiveFee)
            let data = this.composeTx(txParams);
            let { txSize, mass } = data.tx.getMassAndSize();
            let dataFee = this.minimumRequiredTransactionRelayFee(mass);
            const priorityFee = txParamsArg.fee;
            if (txParamsArg.compoundingUTXO) {
                inclusiveFee = true;
                calculateNetworkFee = true;
                txParamsArg.amount = data.amount;
                txParams.amount = data.amount;
                txParams.compoundingUTXO = false;
            }
            const txAmount = txParamsArg.amount;
            let amountRequested = txParamsArg.amount + priorityFee;
            let amountAvailable = data.utxos.map(utxo => utxo.satoshis).reduce((a, b) => a + b, 0);
            this.logger.verbose(`tx ... need data fee: ${PUG(dataFee)} total needed: ${PUG(amountRequested + dataFee)}`);
            this.logger.verbose(`tx ... available: ${PUG(amountAvailable)} in ${data.utxos.length} UTXOs`);
            if (networkFeeMax && dataFee > networkFeeMax) {
                throw new Error(`Fee max is ${networkFeeMax} but the minimum fee required for this transaction is ${PUG(dataFee)} PUG`);
            }
            if (calculateNetworkFee) {
                do {
                    //console.log(`insufficient data fees... incrementing by ${dataFee}`);
                    txParams.fee = priorityFee + dataFee;
                    if (inclusiveFee) {
                        txParams.amount = txAmount - txParams.fee;
                    }
                    this.logger.verbose(`tx ... insufficient data fee for transaction size of ${txSize} bytes`);
                    this.logger.verbose(`tx ... need data fee: ${PUG(dataFee)} for ${data.utxos.length} UTXOs`);
                    this.logger.verbose(`tx ... rebuilding transaction with additional inputs`);
                    let utxoLen = data.utxos.length;
                    this.logger.debug(`final fee ${txParams.fee}`);
                    data = this.composeTx(txParams);
                    ({ txSize, mass } = data.tx.getMassAndSize());
                    dataFee = this.minimumRequiredTransactionRelayFee(mass);
                    if (data.utxos.length != utxoLen)
                        this.logger.verbose(`tx ... aggregating: ${data.utxos.length} UTXOs`);
                } while ((!networkFeeMax || txParams.fee <= networkFeeMax) && txParams.fee < dataFee + priorityFee);
                if (networkFeeMax && txParams.fee > networkFeeMax)
                    throw new Error(`Maximum network fee exceeded; need: ${PUG(dataFee)} PUG maximum is: ${PUG(networkFeeMax)} PUG`);
            }
            else if (dataFee > priorityFee) {
                throw new Error(`Minimum fee required for this transaction is ${PUG(dataFee)} PUG`);
            }
            else if (inclusiveFee) {
                txParams.amount -= txParams.fee;
                data = this.composeTx(txParams);
            }
            data.dataFee = dataFee;
            data.totalAmount = txParams.fee + txParams.amount;
            data.txSize = txSize;
            data.note = txParamsArg.note || "";
            return data;
        });
    }
    /**
     * Build a transaction. Returns transaction info.
     * @param txParams
     * @param txParams.toAddr To address in cashaddr format (e.g. pugdagtest:qq0d6h0prjm5mpdld5pncst3adu0yam6xch4tr69k2)
     * @param txParams.amount Amount to send in sompis (100000000 (1e8) sompis in 1 PUG)
     * @param txParams.fee Fee for miners in sompis
     * @throws `FetchError` if endpoint is down. API error message if tx error. Error if amount is too large to be represented as a javascript number.
     */
    buildTransaction(txParamsArg, debug = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const ts0 = Date.now();
            txParamsArg.skipSign = true;
            txParamsArg.privKeysInfo = true;
            const data = yield this.composeTxAndNetworkFeeInfo(txParamsArg);
            const { id, tx, utxos, utxoIds, amount, toAddr, fee, dataFee, totalAmount, txSize, note, privKeys } = data;
            const ts_0 = Date.now();
            tx.sign(privKeys, pugdagcore.crypto.Signature.SIGHASH_ALL, 'schnorr');
            const { mass: txMass } = tx.getMassAndSize();
            this.logger.info("txMass", txMass);
            if (txMass > Wallet.MaxMassAcceptedByBlock) {
                throw new Error(`Transaction size/mass limit reached. Please reduce this transaction amount. (Mass: ${txMass})`);
            }
            const ts_1 = Date.now();
            //const rawTx = tx.toString();
            const ts_2 = Date.now();
            this.logger.info(`tx ... required data fee: ${PUG(dataFee)} (${utxos.length} UTXOs)`); // (${PUG(txParamsArg.fee)}+${PUG(dataFee)})`);
            //this.logger.verbose(`tx ... final fee: ${PUG(dataFee+txParamsArg.fee)} (${PUG(txParamsArg.fee)}+${PUG(dataFee)})`);
            this.logger.info(`tx ... resulting total: ${PUG(totalAmount)}`);
            //console.log(utxos);
            if (debug || this.loggerLevel > 0) {
                this.logger.debug("submitTransaction: estimateTx", data);
                this.logger.debug("sendTx:utxos", utxos);
                this.logger.debug("::utxos[0].script::", utxos[0].script);
                //console.log("::utxos[0].address::", utxos[0].address)
            }
            const { nLockTime: lockTime, version } = tx;
            if (debug || this.loggerLevel > 0)
                this.logger.debug("composeTx:tx", "txSize:", txSize);
            const ts_3 = Date.now();
            const inputs = tx.inputs.map((input) => {
                if (debug || this.loggerLevel > 0) {
                    this.logger.debug("input.script.inspect", input.script.inspect());
                }
                return {
                    previousOutpoint: {
                        transactionId: input.prevTxId.toString("hex"),
                        index: input.outputIndex
                    },
                    signatureScript: input.script.toBuffer().toString("hex"),
                    sequence: input.sequenceNumber,
                    sigOpCount: 1
                };
            });
            const ts_4 = Date.now();
            const outputs = tx.outputs.map((output) => {
                return {
                    amount: output.satoshis,
                    scriptPublicKey: {
                        scriptPublicKey: output.script.toBuffer().toString("hex"),
                        version: 0
                    }
                };
            });
            const ts_5 = Date.now();
            //const payloadStr = "0000000000000000000000000000000";
            //const payload = Buffer.from(payloadStr).toString("base64");
            //console.log("payload-hex:", Buffer.from(payloadStr).toString("hex"))
            //@ ts-ignore
            //const payloadHash = pugdagcore.crypto.Hash.sha256sha256(Buffer.from(payloadStr));
            const rpcTX = {
                transaction: {
                    version,
                    inputs,
                    outputs,
                    lockTime,
                    //payload:'f00f00000000000000001976a914784bf4c2562f38fe0c49d1e0538cee4410d37e0988ac',
                    payloadHash: '0000000000000000000000000000000000000000000000000000000000000000',
                    //payloadHash:'afe7fc6fe3288e79f9a0c05c22c1ead2aae29b6da0199d7b43628c2588e296f9',
                    //
                    subnetworkId: this.subnetworkId, //Buffer.from(this.subnetworkId, "hex").toString("base64"),
                    fee,
                    //gas: 0
                }
            };
            //const rpctx = JSON.stringify(rpcTX, null, "  ");
            const ts1 = Date.now();
            this.logger.info(`tx ... generation time ${((ts1 - ts0) / 1000).toFixed(2)} sec`);
            if (debug || this.loggerLevel > 0) {
                this.logger.debug(`rpcTX ${JSON.stringify(rpcTX, null, "  ")}`);
                this.logger.debug(`rpcTX ${JSON.stringify(rpcTX)}`);
            }
            const ts_6 = Date.now();
            this.logger.info(`time in msec`, {
                "total": ts_6 - ts0,
                "estimate-transaction": ts_0 - ts0,
                "tx.sign": ts_1 - ts_0,
                "tx.toString": ts_2 - ts_1,
                //"ts_3-ts_2": ts_3-ts_2,
                "tx.inputs.map": ts_4 - ts_3,
                "tx.outputs.map": ts_5 - ts_4,
                //"ts_6-ts_5": ts_6-ts_5
            });
            if (txParamsArg.skipUTXOInUseMark !== true) {
                this.utxoSet.updateUsed(utxos);
            }
            //const rpctx = JSON.stringify(rpcTX, null, "  ");
            //console.log("rpcTX", rpcTX)
            //console.log("\n\n########rpctx\n", rpctx+"\n\n\n")
            //if(amount/1e8 > 3)
            //	throw new Error("TODO XXXXXX")
            return Object.assign(Object.assign({}, data), { rpcTX });
        });
    }
    /**
     * Send a transaction. Returns transaction id.
     * @param txParams
     * @param txParams.toAddr To address in cashaddr format (e.g. pugdagtest:qq0d6h0prjm5mpdld5pncst3adu0yam6xch4tr69k2)
     * @param txParams.amount Amount to send in sompis (100000000 (1e8) sompis in 1 PUG)
     * @param txParams.fee Fee for miners in sompis
     * @throws `FetchError` if endpoint is down. API error message if tx error. Error if amount is too large to be represented as a javascript number.
     */
    submitTransaction(txParamsArg, debug = false) {
        return __awaiter(this, void 0, void 0, function* () {
            txParamsArg.skipUTXOInUseMark = true;
            let reverseChangeAddress = false;
            if (!txParamsArg.changeAddrOverride) {
                txParamsArg.changeAddrOverride = this.addressManager.changeAddress.next();
                reverseChangeAddress = true;
            }
            const { rpcTX, utxoIds, amount, toAddr, note, utxos } = yield this.buildTransaction(txParamsArg, debug);
            //console.log("rpcTX:", rpcTX)
            //throw new Error("TODO : XXXX")
            try {
                const ts = Date.now();
                let txid = yield this.api.submitTransaction(rpcTX);
                const ts3 = Date.now();
                this.logger.info(`tx ... submission time ${((ts3 - ts) / 1000).toFixed(2)} sec`);
                this.logger.info(`txid: ${txid}`);
                if (!txid) {
                    if (reverseChangeAddress)
                        this.addressManager.changeAddress.reverse();
                    return null; // as TxResp;
                }
                this.utxoSet.inUse.push(...utxoIds);
                this.txStore.add({
                    in: false, ts, id: txid, amount, address: toAddr, note,
                    blueScore: this.blueScore,
                    tx: rpcTX.transaction,
                    myAddress: this.addressManager.isOur(toAddr),
                    isCoinbase: false,
                    version: 2
                });
                this.updateDebugInfo();
                this.emitCache();
                /*
                this.pendingInfo.add(txid, {
                    rawTx: tx.toString(),
                    utxoIds,
                    amount,
                    to: toAddr,
                    fee
                });
                */
                const resp = {
                    txid,
                    //rpctx
                };
                return resp;
            }
            catch (e) {
                if (reverseChangeAddress)
                    this.addressManager.changeAddress.reverse();
                if (typeof e.setExtraDebugInfo == "function") {
                    let mass = 0;
                    let satoshis = 0;
                    let list = utxos.map(tx => {
                        var _d;
                        mass += tx.mass;
                        satoshis += tx.satoshis;
                        return Object.assign({}, tx, {
                            address: tx.address.toString(),
                            script: (_d = tx.script) === null || _d === void 0 ? void 0 : _d.toString()
                        });
                    });
                    //86500,00000000
                    let info = {
                        mass,
                        satoshis,
                        utxoCount: list.length,
                        utxos: list
                    };
                    e.setExtraDebugInfo(info);
                }
                throw e;
            }
        });
    }
    /*
    * Compound UTXOs by re-sending funds to itself
    */
    compoundUTXOs(txCompoundOptions = {}, debug = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const { UTXOMaxCount = COMPOUND_UTXO_MAX_COUNT, networkFeeMax = 0, fee = 0, useLatestChangeAddress = false } = txCompoundOptions;
            //let toAddr = this.addressManager.changeAddress.next()
            let toAddr = this.addressManager.changeAddress.atIndex[0];
            //console.log("compoundUTXOs: to address:", toAddr, "useLatestChangeAddress:"+useLatestChangeAddress)
            if (useLatestChangeAddress) {
                toAddr = this.addressManager.changeAddress.current.address;
            }
            if (!toAddr) {
                toAddr = this.addressManager.changeAddress.next();
            }
            let txParamsArg = {
                toAddr,
                changeAddrOverride: toAddr,
                amount: -1,
                fee,
                networkFeeMax,
                compoundingUTXO: true,
                compoundingUTXOMaxCount: UTXOMaxCount
            };
            try {
                let res = yield this.submitTransaction(txParamsArg, debug);
                if (!(res === null || res === void 0 ? void 0 : res.txid))
                    this.addressManager.changeAddress.reverse();
                return res;
            }
            catch (e) {
                this.addressManager.changeAddress.reverse();
                throw e;
            }
        });
    }
    /*
    undoPendingTx(id: string): void {
        const {	utxoIds	} = this.pendingInfo.transactions[id];
        delete this.pendingInfo.transactions[id];
        this.utxoSet.release(utxoIds);
        this.addressManager.changeAddress.reverse();
        this.runStateChangeHooks();
    }
    */
    /**
     * After we see the transaction in the API results, delete it from our pending list.
     * @param id The tx hash
     */
    /*
   deletePendingTx(id: string): void {
       // undo + delete old utxos
       const {	utxoIds } = this.pendingInfo.transactions[id];
       delete this.pendingInfo.transactions[id];
       this.utxoSet.remove(utxoIds);
   }
   */
    runStateChangeHooks() {
        //this.utxoSet.updateUtxoBalance();
        //this.updateBalance();
    }
    startUTXOsPolling() {
        //if (this.UTXOsPollingStarted)
        //	return
        //this.UTXOsPollingStarted = true;
        this.emitUTXOs();
    }
    emitUTXOs() {
        let chunks = helper.chunks([...this.utxoSet.utxos.confirmed.values()], 100);
        chunks = chunks.concat(helper.chunks([...this.utxoSet.utxos.pending.values()], 100));
        let send = () => {
            let utxos = chunks.pop();
            if (!utxos)
                return;
            utxos = utxos.map(tx => {
                return Object.assign({}, tx, {
                    address: tx.address.toString()
                });
            });
            this.emit("utxo-sync", { utxos });
            helper.dpc(200, send);
        };
        send();
    }
    get cache() {
        return {
            //pendingTx: this.pendingInfo.transactions,
            utxos: {
                //utxoStorage: this.utxoSet.utxoStorage,
                inUse: this.utxoSet.inUse,
            },
            //transactionsStorage: this.transactionsStorage,
            addresses: {
                receiveCounter: this.addressManager.receiveAddress.counter,
                changeCounter: this.addressManager.changeAddress.counter,
            }
        };
    }
    restoreCache(cache) {
        //this.pendingInfo.transactions = cache.pendingTx;
        //this.utxoSet.utxoStorage = cache.utxos.utxoStorage;
        this.utxoSet.inUse = cache.utxos.inUse;
        /*
        Object.entries(this.utxoSet.utxoStorage).forEach(([addr, utxos]: [string, Api.Utxo[]]) => {
            this.utxoSet.add(utxos, addr);
        });
        this.transactionsStorage = cache.transactionsStorage;
        this.addressManager.getAddresses(cache.addresses.receiveCounter + 1, 'receive');
        this.addressManager.getAddresses(cache.addresses.changeCounter + 1, 'change');
        this.addressManager.receiveAddress.advance(cache.addresses.receiveCounter - 1);
        this.addressManager.changeAddress.advance(cache.addresses.changeCounter);
        //this.transactions = txParser(this.transactionsStorage, Object.keys(this.addressManager.all));
        this.runStateChangeHooks();
        */
    }
    /**
     * Generates encrypted wallet data.
     * @param password user's chosen password
     * @returns Promise that resolves to object-like string. Suggested to store as string for .import().
     */
    export(password) {
        return __awaiter(this, void 0, void 0, function* () {
            const savedWallet = {
                privKey: this.HDWallet.toString(),
                seedPhrase: this.mnemonic
            };
            return crypto_1.Crypto.encrypt(password, JSON.stringify(savedWallet));
        });
    }
    setLogLevel(level) {
        this.logger.setLevel(level);
        this.loggerLevel = level != 'none' ? 2 : 0;
        pugdagcore.setDebugLevel(level ? 1 : 0);
    }
}
exports.Wallet = Wallet;
_a = BALANCE_CONFIRMED, _b = BALANCE_PENDING, _c = BALANCE_TOTAL;
Wallet.Mnemonic = Mnemonic;
Wallet.passwordHandler = crypto_1.Crypto;
Wallet.Crypto = crypto_1.Crypto;
Wallet.pugdagcore = pugdagcore;
Wallet.COMPOUND_UTXO_MAX_COUNT = COMPOUND_UTXO_MAX_COUNT;
Wallet.MaxMassAcceptedByBlock = 100000;
Wallet.MaxMassUTXOs = 100000;
//Wallet.MaxMassAcceptedByBlock -
//pugdagcore.Transaction.EstimatedStandaloneMassWithoutInputs;
// TODO - integrate with Pugdagcore-lib
Wallet.networkTypes = {
    pugdag: { port: 26589, network: 'pugdag', name: 'mainnet' },
    pugdagtest: { port: 26689, network: 'pugdagtest', name: 'testnet' },
    pugdagsim: { port: 26789, network: 'pugdagsim', name: 'simnet' },
    pugdagdev: { port: 26889, network: 'pugdagdev', name: 'devnet' }
};
Wallet.networkAliases = {
    mainnet: 'pugdag',
    testnet: 'pugdagtest',
    devnet: 'pugdagdev',
    simnet: 'pugdagsim'
};
