// @pugdag/wallet/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pugdagcore = exports.FlowLogger = exports.Storage = exports.helper = exports.EventTargetImpl = exports.log = exports.initPugdagFramework = exports.Wallet = exports.COINBASE_CFM_COUNT = exports.CONFIRMATION_COUNT = void 0;
const logger_1 = require("./utils/logger");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return logger_1.log; } });
Object.defineProperty(exports, "FlowLogger", { enumerable: true, get: function () { return logger_1.FlowLogger; } });
const wallet_1 = require("./wallet/wallet");
Object.defineProperty(exports, "Wallet", { enumerable: true, get: function () { return wallet_1.Wallet; } });
Object.defineProperty(exports, "Storage", { enumerable: true, get: function () { return wallet_1.Storage; } });
Object.defineProperty(exports, "pugdagcore", { enumerable: true, get: function () { return wallet_1.pugdagcore; } });
Object.defineProperty(exports, "CONFIRMATION_COUNT", { enumerable: true, get: function () { return wallet_1.CONFIRMATION_COUNT; } });
Object.defineProperty(exports, "COINBASE_CFM_COUNT", { enumerable: true, get: function () { return wallet_1.COINBASE_CFM_COUNT; } });
const initPugdagFramework_1 = require("./wallet/initPugdagFramework");
Object.defineProperty(exports, "initPugdagFramework", { enumerable: true, get: function () { return initPugdagFramework_1.initPugdagFramework; } });
const event_target_impl_1 = require("./wallet/event-target-impl");
Object.defineProperty(exports, "EventTargetImpl", { enumerable: true, get: function () { return event_target_impl_1.EventTargetImpl; } });
const helper = __importStar(require("./utils/helper"));
exports.helper = helper;
